src/v/io/io_queue.h

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "io/page.h"
+#include "io/persistence.h"
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/semaphore.hh>
+
+namespace experimental::io {
+
+namespace testing_details {
+class io_queue_accessor;
+}; // namespace testing_details
+
+/**
+ * Manager of I/O requests for a single file.
+ *
+ * An I/O request is added to the queue using the `submit_*` interfaces. For
+ * example, a page may be read or written by passing a reference to the page to
+ * either `submit_read` or `submit_write`.
+ *
+ * There is no guarantee of the order in which I/O requests are processed.
+ *
+ * Submission of I/O requests can be made at any time, but pending I/O requests
+ * are processed only when the queue is in the open state. A new I/O queue is in
+ * the closed state. To open the queue use the `open()` interface which will
+ * open a handle to the underlying file. To close the queue and the file handle
+ * use the `close()` interface. A queue may be opened and closed any number of
+ * times. This is useful if the queue is used in a context that control the
+ * maximum number of open file handles or wants to pause I/O for some reason.
+ */
+class io_queue {
+public:
+    /**
+     * Callback invoked when the scheduler completes an operation.
+     */
+    using completion_callback_type = seastar::noncopyable_function<void(page&)>;
+
+    /**
+     * Construct a new io_queue.
+     */
+    io_queue(
+      persistence* storage,
+      std::filesystem::path path,
+      completion_callback_type complete) noexcept;
+
+    io_queue(const io_queue&) = delete;
+    io_queue& operator=(const io_queue&) = delete;
+    io_queue(io_queue&&) noexcept = delete;
+    io_queue& operator=(io_queue&&) noexcept = delete;
+    ~io_queue() noexcept = default;
+
+    /**
+     * Starts the I/O queue's background dispatcher.
+     */
+    void start() noexcept;
+
+    /**
+     * Stops the I/O queue's background dispatcher.
+     */
+    seastar::future<> stop() noexcept;
+
+    /**
+     * Return the path of the backing file.
+     */
+    [[nodiscard]] const std::filesystem::path& path() const noexcept;
+
+    /**
+     * Open the queue and begin dispatching pending I/O requests.
+     *
+     * Requires that opened() be false.
+     */
+    seastar::future<> open() noexcept;
```

Comment:
The `noexcept` for me is always a little tricky because this can return a failed future still, meaning that `opened` is still false... I don't have a suggestion outside documenting that this can fail?

Reply:
I think this is self-documenting, right? The future can contain an exception, but open() will never let a raw exception leak out. This is the pattern used for most of the seastar I/O layer IIRC.

Reply:
I mean I guess so, part of it is that if open returns an error it's OK to be retried. So folks still have to handle potential failures in `close`?

Reply:
added this comment

```
diff --git a/src/v/io/io_queue.h b/src/v/io/io_queue.h
index d4d916c8a7..baf819e9b9 100644
--- a/src/v/io/io_queue.h
+++ b/src/v/io/io_queue.h
@@ -79,7 +79,8 @@ public:
     /**
      * Open the queue and begin dispatching pending I/O requests.
      *
-     * Requires that opened() be false.
+     * Requires that opened() be false. The queue will remain in the closed
+     * state if an exceptional future is returned.
      */
     seastar::future<> open() noexcept;
```

Reply:
That is helpful IMO.